src/Microsoft.OpenApi.Readers/V2/OpenApiDocumentDeserializer.cs

```diff
@@ -119,21 +119,76 @@ internal static partial class OpenApiV2Deserializer
             {s => s.StartsWith("x-"), (o, p, n) => o.AddExtension(p, n.CreateAny())}
         };
 
-        private static void MakeServers(IList<OpenApiServer> servers, ParsingContext context)
+        private static void MakeServers(IList<OpenApiServer> servers, ParsingContext context, Uri defaultUrl)
         {
             var host = context.GetFromTempStorage<string>("host");
             var basePath = context.GetFromTempStorage<string>("basePath");
             var schemes = context.GetFromTempStorage<List<string>>("schemes");
 
+            // If nothing is provided, don't create a server
+            if (host == null && basePath == null && schemes == null)
+            {
+                return;
+            }
+
+            // Fill in missing information based on the defaultUrl
+            if (defaultUrl != null)
+            {
+                host = host ?? defaultUrl.GetComponents(UriComponents.NormalizedHost, UriFormat.SafeUnescaped);
+                basePath = basePath ?? defaultUrl.GetComponents(UriComponents.Path, UriFormat.SafeUnescaped);
+                schemes = schemes ?? new List<string> { defaultUrl.GetComponents(UriComponents.Scheme, UriFormat.SafeUnescaped) };
+            } else if (String.IsNullOrEmpty(host) && String.IsNullOrEmpty(basePath))
+            {
+                return;  // Can't make a server object out of just a Scheme
+            }
+            
+            // Create the Server objects
             if (schemes != null)
             {
                 foreach (var scheme in schemes)
                 {
-                    var server = new OpenApiServer();
-                    server.Url = scheme + "://" + (host ?? "example.org/") + (basePath ?? "/");
+                    var uriBuilder = new UriBuilder(scheme, host)
+                    {
+                        Path = basePath
+                    };
+
+                    var server = new OpenApiServer
+                    {
+                        Url = uriBuilder.ToString()
+                    };
+
                     servers.Add(server);
                 }
             }
+            else
+            {
+                if (!String.IsNullOrEmpty(host))
+                {
+                    host = "//" + host;
```

Comment:
>= "//" + host; [](start = 25, length = 14)

nit, it wasn't obvious to me in the beginning that we need this "//" to denote "the current protocol". 

Could you add some comments to explain this?